File: FantasyRadio/src/main/java/ru/sigil/fantasyradio/utils/DownloadThread.java
package ru.sigil.fantasyradio.utils;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.widget.Toast;

import javax.inject.Inject;

import ru.sigil.fantasyradio.dagger.Bootstrap;
import ru.sigil.fantasyradio.saved.MP3Entity;

public class DownloadThread extends Thread {
    private String fileUrl;
    private String fileDir;
    private String fileName;
    private Context context;
    private Handler finishedHandler;
    private MP3Entity mp3Entity;
    private Handler errorHandler;

    Object lock = new Object();
    private final static String[] ReservedChars = {"|", "\\", "?", "*", "<", "\"",
            ":", ">", "+", "[", "]", "/", "'", "%"};

    @Inject
    FileDownloader fileDownloader;

    public DownloadThread(String fileUrl, String fileDir, String fileName,
                          Context c, Handler h, MP3Entity entity, Handler errorHandler) {
        this.fileDir = fileDir;
        this.fileUrl = fileUrl;
        this.fileName = fileName;
        this.context = c;
        this.finishedHandler = h;
        this.errorHandler = errorHandler;
        mp3Entity = entity;
        Bootstrap.INSTANCE.getBootstrap().inject(this);
    }

    @Override
    public void run() {
        for (String s : ReservedChars) {
            fileName = fileName.replace(s, "_");
        }
        if(fileDownloader.DownloadFile(fileUrl, fileDir, fileName, context)) {
            mp3Entity.setDirectory(fileDir + fileName);
            Message msg = new Message();
            Bundle b = new Bundle();
            b.putString("artist", fileName.substring(0, 11));
            b.putString("title", mp3Entity.getTitle());
            b.putString("directory", mp3Entity.getDirectory());
            b.putString("time", mp3Entity.getTime());
            // --------------------------------
            b.putString("URL", fileUrl);
            // --------------------------------
            msg.setData(b);
            finishedHandler.sendMessage(msg);
        }
        else
        {
            errorHandler.sendEmptyMessage(0);
        }
    }
}

File: FantasyRadio/src/main/java/ru/sigil/fantasyradio/utils/DownloadThread.kt
package ru.sigil.fantasyradio.utils

import android.os.Handler
import ru.sigil.fantasyradio.saved.MP3Entity
import android.os.Bundle
import android.os.Message

/**
 * Created by namelessone
 * on 08.12.18.
 */
class DownloadThread(private val fileDownloader: IFileDownloader, private val fileUrl: String?,
                     private val fileDir: String, private var fileName: String,
                     private val finishedHandler: Handler, private val mp3Entity: MP3Entity,
                     private val errorHandler: Handler) : Thread() {
    private val reservedChars = arrayOf("|", "\\", "?", "*", "<", "\"", ":", ">", "+", "[", "]", "/", "'", "%")

    override fun run() {
        for (s in reservedChars) {
            fileName = fileName.replace(s, "_")
        }
        if (fileDownloader.downloadFile(fileUrl ?: "", fileDir, fileName)) {
            mp3Entity.directory = fileDir + fileName
            val msg = Message()
            val b = Bundle()
            b.putString("artist", fileName.substring(0, 11))
            b.putString("title", mp3Entity.title)
            b.putString("directory", mp3Entity.directory)
            b.putString("time", mp3Entity.time)
            // --------------------------------
            b.putString("URL", fileUrl)
            // --------------------------------
            msg.data = b
            finishedHandler.sendMessage(msg)
        } else {
            errorHandler.sendEmptyMessage(0)
        }
    }
}